peak_finder/src/peak_finder_component.cpp

```diff
@@ -0,0 +1,218 @@
+      std::bind(
+        &PeakFinderComponent::handle_goal, this, std::placeholders::_1,
+        std::placeholders::_2),
+      std::bind(&PeakFinderComponent::handle_cancel, this, std::placeholders::_1),
+      std::bind(&PeakFinderComponent::handle_accepted, this, std::placeholders::_1));
+
+    elevation_client_ = create_client<stsl_interfaces::srv::SampleElevation>("/sample_elevation");
+  }
+
+private:
+  tf2_ros::Buffer tf_buffer_;
+  tf2_ros::TransformListener tf_listener_;
+  rclcpp_action::Server<ParkAtPeak>::SharedPtr action_server_;
+  rclcpp::Client<stsl_interfaces::srv::SampleElevation>::SharedPtr elevation_client_;
+  Navigator navigator_;
+
+  rclcpp_action::GoalResponse handle_goal(
+    const rclcpp_action::GoalUUID & /*uuid*/,
+    std::shared_ptr<const ParkAtPeak::Goal>/*goal*/)
+  {
+    return rclcpp_action::GoalResponse::ACCEPT_AND_EXECUTE;
+  }
+
+  rclcpp_action::CancelResponse
+  handle_cancel(const std::shared_ptr<ParkAtPeakGoalHandle> goal_handle)
+  {
+    return rclcpp_action::CancelResponse::ACCEPT;
+  }
+
+  void handle_accepted(const std::shared_ptr<ParkAtPeakGoalHandle> goal_handle)
+  {
+    std::thread{std::bind(&PeakFinderComponent::execute, this, std::placeholders::_1),
+      goal_handle}
+    .detach();
+  }
+
+  void execute(const std::shared_ptr<ParkAtPeakGoalHandle> goal_handle)
+  {
+    if (!elevation_client_->service_is_ready()) {
+      RCLCPP_ERROR(
+        get_logger(), "%s service must be available to run peak_finder action!",
+        elevation_client_->get_service_name());
+      goal_handle->abort(std::make_shared<ParkAtPeak::Result>());
+      return;
+    }
+
+    std::string tf_error_msg;
+    if (!tf_buffer_.canTransform("map", "base_footprint", tf2::TimePointZero, &tf_error_msg)) {
+      RCLCPP_ERROR(
+        get_logger(), "Robot position could not be looked up via TF. Error: %s",
+        tf_error_msg.c_str());
+      goal_handle->abort(std::make_shared<ParkAtPeak::Result>());
+      return;
+    }
+
+    if (!navigator_.server_available()) {
+      RCLCPP_ERROR(
+        get_logger(),
+        "/navigate_to_point action must be available to run peak_finder action!");
+      goal_handle->abort(std::make_shared<ParkAtPeak::Result>());
+      return;
+    }
+
+    try {
+      while (rclcpp::ok() && !goal_handle->is_canceling()) {
+        const auto robot_position = GetRobotPosition();
+
+        const double current_elevation = SampleElevation(robot_position);
+
+        double search_radius;
+        get_parameter("search_radius", search_radius);
+        int sample_count;
+        get_parameter("sample_count", sample_count);
+
+        auto generate_next_pose =
+          [&robot_position, search_radius, sample_count, angle = 0.0]() mutable {
+            Eigen::Vector2d pose =
+              (search_radius * Eigen::Vector2d(std::cos(angle), std::sin(angle))) + robot_position;
+            angle += (2 * M_PI) / sample_count;
+            return pose;
+          };
```

Comment:
Why lambda this instead of making a separate function?

Reply:
Mostly because it needs to capture local variables and maintain its own state ( `angle` ).